transport/cluster.go

```diff
@@ -0,0 +1,252 @@
+			node.tokens = v[nodeTokens]
+			m[addr] = node
+		} else {
+			m[addr] = &Node{
+				addr:       addr,
+				datacenter: string(v[nodeDC]),
+				rack:       string(v[nodeRack]),
+				tokens:     v[nodeTokens],
+				pool:       InitNodeConnPool(appendDefaultPort(addr), c.cfg),
+				status:     atomic.NewBool(statusUP),
+			}
+		}
+	}
+
+	// TODO: add closing conn pools from deleted nodes.
+	c.SetPeers(m)
+	return nil
+}
+
+var (
+	peerQuery = Statement{
+		Content:     "SELECT peer, data_center, rack, tokens FROM system.peers",
+		Consistency: frame.ONE,
+	}
+
+	localQuery = Statement{
+		Content:     "SELECT rpc_address, data_center, rack, tokens FROM system.local",
+		Consistency: frame.ONE,
+	}
+)
+
+// Indexes of columns returned from refresh topology queries.
+const (
+	nodeAddr   = 0
+	nodeDC     = 1
+	nodeRack   = 2
+	nodeTokens = 3
+)
+
+func (c *Cluster) getAllNodesInfo() ([]frame.Row, error) {
+	peerRes, err := c.control.Query(peerQuery, nil)
+	if err != nil {
+		return nil, fmt.Errorf("discovering peer topology: %w", err)
+	}
+
+	localRes, err := c.control.Query(localQuery, nil)
+	if err != nil {
+		return nil, fmt.Errorf("discovering local topology: %w", err)
+	}
+
+	rows := append(peerRes.Rows, localRes.Rows[0])
+	return rows, nil
+}
+
+func (c *Cluster) GetPeers() PeerMap {
+	return c.peers.Load().(PeerMap)
+}
+
+func (c *Cluster) SetPeers(m PeerMap) {
+	c.peers.Store(m)
+}
+
+const defaultPort = 9042
+
+func appendDefaultPort(addr string) string {
+	return addr + ":" + strconv.Itoa(defaultPort)
+}
+
+const eventChanSize = 32
+
+func (c *Cluster) handleEvents() {
+	c.events = make(responseHandler, eventChanSize)
+	for {
+		res, ok := <-c.events
+		if !ok {
+			return
+		}
+		switch v := res.Response.(type) {
+		case *TopologyChange:
+			c.handleTopologyChange()
+		case *StatusChange:
+			c.handleStatusChange(v)
+		case *SchemaChange:
+			// TODO: add schema change.
+		default:
+			log.Printf("unsupported event type")
+		}
+	}
+}
+
+func (c *Cluster) handleTopologyChange() {
+	c.RequestRefresh()
+}
+
+func (c *Cluster) handleStatusChange(v *StatusChange) {
+	m := c.GetPeers()
+	addr := v.Address.String()
+	if n, ok := m[addr]; ok {
+		n.SetStatus(v.Status)
+	} else {
+		log.Printf("node which status is being set to %s was not present in known topology: %s", v.Status, addr)
+		c.RequestRefresh()
+	}
+}
+
+// RequestRefresh notifies cluster that it should update topology.
+// TODO: do we need some mechanism for notifying requester that refresh was successful?
+func (c *Cluster) RequestRefresh() {
+	c.refresher <- struct{}{}
+}
+
+func (c *Cluster) StopEvenHandling() {
+	close(c.events)
+}
+
+const (
+	refreshChanSize = 32
+	refreshInterval = 60 * time.Second
+)
+
+// Loop handles refresh topology requests.
+func (c *Cluster) Loop() {
+	c.refresher = make(refreshHandler, refreshChanSize)
+	ticker := time.NewTimer(refreshInterval)
+	for {
+		select {
+		case _, ok := <-c.refresher:
+			if ok {
+				c.tryRefresh()
+				ticker.Stop()
```

Comment:
Should ticker be stopped right before its reset?

Reply:
Documentation: "For a Timer created with NewTimer, Reset should be invoked only on stopped or expired timers with drained channels".